File: backend/app/schemas/kb.py
"""知识库域请求/响应模型。"""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


# ==================== 知识库 ====================
class KbCreate(BaseModel):
    """创建知识库请求。"""

    name: str = Field(..., min_length=1, max_length=128, description="知识库名称")
    description: str | None = Field(None, max_length=2000, description="知识库描述")
    category: str | None = Field(None, max_length=64, description="类目标签(如 3C数码)")


class KbUpdate(BaseModel):
    """更新知识库请求(部分字段可选)。"""

    name: str | None = Field(None, min_length=1, max_length=128)
    description: str | None = Field(None, max_length=2000)
    category: str | None = Field(None, max_length=64)


class KbOut(BaseModel):
    """知识库信息(含冗余统计)。"""

    id: int
    name: str
    description: str | None
    category: str | None
    doc_count: int
    chunk_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ==================== 文档 ====================
class DocumentOut(BaseModel):
    """文档记录(含解析任务状态,前端据此轮询进度)。"""

    id: int
    kb_id: int
    filename: str
    file_type: str
    size_bytes: int
    parse_status: str            # pending / parsing / done / failed
    progress_done: int
    progress_total: int
    error_msg: str | None
    version: int
    chunk_count: int
    sha256: str
    created_at: datetime
    finished_at: datetime | None

    model_config = {"from_attributes": True}


# ==================== 分块 ====================
class ChunkOut(BaseModel):
    """分块信息(管理端浏览/搜索)。"""

    id: str
    doc_id: int
    kb_id: int
    seq: int
    text: str
    meta_json: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ==================== 检索调试 ====================
class DebugSearchRequest(BaseModel):
    """调试台请求:输入问题,输出各检索阶段明细。"""

    question: str = Field(..., min_length=1, description="测试问题")
    kb_id: int | None = Field(None, description="限定知识库(为空则全库检索)")


class DebugHit(BaseModel):
    """调试台单条命中(带来源与分数)。"""

    chunk_id: str
    score: float
    source: str = Field(default="", description="来源文件名")
    text: str = Field(default="", description="片段文本(预览节选)")


class DebugSearchResult(BaseModel):
    """调试台响应:展示 向量路 / 关键词路 / RRF 融合 / 重排后 各阶段结果。"""

    vector_hits: list[DebugHit]
    keyword_hits: list[DebugHit]
    fused_hits: list[DebugHit]
    reranked_hits: list[DebugHit]
    refused: bool = Field(..., description="按当前拒答阈值是否应拒答")
    threshold: float

File: frontend/src/api/kb.ts
import http from './http'

/** 知识库管理相关接口(全部仅管理员可用,后端 403 兜底)。 */

export interface KbInfo {
  id: number
  name: string
  description?: string | null
  category?: string | null
  doc_count: number
  chunk_count: number
  created_at: string
  updated_at: string
}

export interface DocumentItem {
  id: number
  kb_id: number
  filename: string
  file_type: string
  size_bytes: number
  parse_status: 'pending' | 'parsing' | 'done' | 'failed'
  progress_done: number
  progress_total: number
  error_msg?: string | null
  version: number
  chunk_count: number
  sha256: string
  created_at: string
  finished_at?: string | null
}

export interface ChunkItem {
  id: string
  doc_id: number
  kb_id: number
  seq: number
  text: string
  meta_json?: string | null
  created_at: string
}

export interface Page<T> {
  items: T[]
  total: number
}

export interface DebugHit {
  chunk_id: string
  score: number
  source?: string
  text: string
}

export interface DebugResult {
  vector_hits: DebugHit[]
  keyword_hits: DebugHit[]
  fused_hits: DebugHit[]
  reranked_hits: DebugHit[]
  refused: boolean
  threshold: number
}

// ==================== 知识库 ====================
/** 登录用户(含普通用户)可用的知识库名列表,供问答页选择检索范围 */
export function listKbNames() {
  return http.get<{ id: number; name: string }[]>('/kb/public')
}

export const kbApi = {
  list: () => http.get<KbInfo[]>('/kb'),
  get: (id: number) => http.get<KbInfo>(`/kb/${id}`),
  create: (data: Partial<KbInfo>) => http.post<KbInfo>('/kb', data),
  update: (id: number, data: Partial<KbInfo>) => http.patch<KbInfo>(`/kb/${id}`, data),
  remove: (id: number) => http.delete(`/kb/${id}`),
}

// ==================== 文档 ====================
export const docApi = {
  list: (kbId: number) => http.get<DocumentItem[]>(`/documents/kb/${kbId}`),
  get: (id: number) => http.get<DocumentItem>(`/documents/${id}`),
  remove: (id: number) => http.delete(`/documents/${id}`),
  reparse: (id: number) => http.post<DocumentItem>(`/documents/${id}/reparse`),
  /** 上传单个文件;replace=true 表示已存在时覆盖重传 */
  upload: (kbId: number, file: File, replace = false) => {
    const form = new FormData()
    form.append('file', file)
    return http.post<DocumentItem>(`/documents/kb/${kbId}/upload?replace=${replace}`, form, {
      headers: { 'Content-Type': 'multipart/form-data' },
      timeout: 0, // 大文件上传不限时
    })
  },
}

// ==================== 分块 ====================
export const chunkApi = {
  list: (kbId: number, params: { page?: number; size?: number; q?: string }) =>
    http.get<Page<ChunkItem>>(`/chunks/kb/${kbId}`, { params }),
  remove: (chunkId: string) => http.delete(`/chunks/${chunkId}`),
}

// ==================== 检索调试 ====================
export const debugApi = {
  search: (data: { question: string; kb_id?: number | null }) =>
    http.post<DebugResult>('/debug/search', data),
}
